aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -597,6 +604,44 @@ struct cpu_scatter_gather_base_kernel {
 //
 //   step 2: spmm reduce, parallel on M and vectorize on K
 //
+
+template <typename scalar_t, ReductionType reduce>
+inline void _init(scalar_t* self_ptr, at::opmath_type<scalar_t>* buffer_ptr, int64_t K, bool include_self) {
+  if (!include_self) {
+    init<at::opmath_type<scalar_t>, reduce>(buffer_ptr, K, include_self);
+  } else {
+    vec::convert(self_ptr, buffer_ptr, K);
+  }
+}
+
+template <typename scalar_t, ReductionType reduce>
+inline void _update(scalar_t* src_data, at::opmath_type<scalar_t>* buffer_ptr, int64_t col, int64_t K) {
+  if constexpr (!is_reduced_floating_point_v<scalar_t>) {
```

Comment:
it is preferred to do template specification in `aten/src/ATen/native/cpu/ReduceUtils.h` upon `init` and `update` directly, something like below. Probably this is should be overload, i am not sure, not a pragma guy ...

```C++
// the first template for float32, float64
template <typename scalar_t, ReductionType reduce,
                  typename std::enable_if_t<!is_reduced_floating_point_v<scalar_t>, int> = 0>
inline void update(scalar_t* out, scalar_t* data, int64_t K) {}

// the second template for float16, bfloat16
template <typename scalar_t, ReductionType reduce,
                  typename std::enable_if_t<is_reduced_floating_point_v<scalar_t>, int> = 0>
inline void update(scalar_t* out, float* data, int64_t K) {}
```